assignment4/count_islands.py

```diff
@@ -0,0 +1,128 @@
+'''
+Created on 23 June 2017
+
+@author: Eva
+'''
+
+from disjoint_set import DisjointSet
+
+def count_islands_iter(rowMax, colMax, tiles):
+    """Counts the number of islands that exist in the given array.
+    
+    This approach uses a disjoint set to merge the land tiles that
+    are adjacent either vertically or horizontally to a given land tile.
+    It then counts the roots of trees that represent land tiles to 
+    calculate the islands.
+    
+    NB. True -> Land tile
+        False -> Water tile
+    NB2. Assume rowMax and colMax are not out of bounds of the array
+        
+        Args:
+            rowMax: an integer, the number of rows in the array
+            colMax: an integer, the number of columns in the array
+            tiles: a two-dimensional array of boolean values
+        
+        Returns:
+            The number of islands found in the array.
+    """
+    if tiles is None or len(tiles) == 0 or len(tiles[0]) == 0: 
+        raise ValueError("Please insert a non-empty array of tiles.")
+    
+    dis_set = DisjointSet(rowMax * colMax)
+    
+    for i in range(rowMax):
+        for j in range(colMax):
+            if tiles[i][j] == False: continue
+            
+            # right
+            if in_bound(j+1, colMax) and tiles[i][j+1] == True:
+                dis_set.merge(i*rowMax+j, i*rowMax+j+1)
+            # left
+            if in_bound(j-1, colMax) and tiles[i][j-1] == True:
+                dis_set.merge(i*rowMax+j, i*rowMax+j-1)
+            # top
+            if in_bound(i+1, rowMax) and tiles[i+1][j]:
+                dis_set.merge(i*rowMax+j, (i+1)*rowMax+j)
+            # bottom
+            if in_bound(i-1, rowMax) and tiles[i-1][j]:
+                dis_set.merge(i*rowMax+j, (i-1)*rowMax+j)
+    
+    visited_roots = []
```

Comment:
Optimisation:
The only operations you do on `visited_roots` is adding elements and checking membership. It would seem that this would be more efficient with a `set()` because checking membership is O(1) for sets but linear for the list.